File: buildy_lib/executor.py
"""Execute tasks with caching, parallel execution, and resource-aware scheduling"""

import os
import time
import subprocess
import logging
from typing import Dict, List
from concurrent.futures import ThreadPoolExecutor, as_completed

from .models import BuildTask
from .cache import BuildCache
from .task_graph import TaskGraph
from .execution import ExecutionEnvironment, NativeExecution
from .constants import DEFAULT_TASK_TIMEOUT_SECONDS

logger = logging.getLogger('buildy.executor')

class TaskExecutor:
    """Execute tasks with caching, parallel execution, and resource-aware scheduling"""

    def __init__(self, cache: BuildCache, max_workers: int = 4, max_memory_mb: int = 8192,
                 exec_env: ExecutionEnvironment = None):
        self.cache = cache
        self.max_workers = max_workers
        self.max_memory_mb = max_memory_mb
        self.exec_env = exec_env or NativeExecution()  # Default to native execution
        self.execution_stats = {
            'total_tasks': 0,
            'cache_hits': 0,
            'cache_misses': 0,
            'failed_tasks': 0,
            'total_time': 0
        }

    def execute_task_graph(self, graph: TaskGraph, dry_run: bool = False) -> bool:
        """Execute all tasks in the graph"""
        if not graph.execution_stages:
            graph.build_execution_stages()

        start_time = time.time()
        success = True
        completed_tasks = 0
        total_tasks = len(graph.tasks)

        logger.info(f"Starting build with {total_tasks} tasks in {len(graph.execution_stages)} stages")

        if dry_run:
            self._print_dry_run(graph)
            return True

        for stage_num, stage_tasks in enumerate(graph.execution_stages, 1):
            progress = (completed_tasks / total_tasks) * 100 if total_tasks > 0 else 0
            logger.info(f"\nStage {stage_num}/{len(graph.execution_stages)}: "
                       f"{len(stage_tasks)} task(s) [{progress:.1f}% complete]")

            stage_success = self._execute_stage(stage_tasks, graph.tasks)
            completed_tasks += len(stage_tasks)
            
            if not stage_success:
                success = False
                break

        self.execution_stats['total_time'] = time.time() - start_time
        self._print_summary()

        return success

    def _execute_stage(self, task_ids: List[str], tasks: Dict[str, BuildTask]) -> bool:
        """Execute a single stage of tasks in parallel with resource-aware scheduling"""
        if len(task_ids) == 1:
            # Single task - execute directly
            return self._execute_single_task(tasks[task_ids[0]])

        # Resource-aware parallel execution
        # Calculate total memory requirements for the stage
        total_memory_required = sum(
            tasks[task_id].resource_requirements.memory_mb 
            for task_id in task_ids
        )
        
        # Adjust max workers based on memory constraints
        if total_memory_required > self.max_memory_mb:
            # Need to limit parallelism to avoid memory overload
            effective_workers = max(1, int(self.max_memory_mb / 
                                          max(tasks[tid].resource_requirements.memory_mb 
                                              for tid in task_ids)))
            effective_workers = min(effective_workers, self.max_workers, len(task_ids))
            logger.warning(f"Limiting parallelism to {effective_workers} workers due to memory constraints "
                         f"({total_memory_required}MB required, {self.max_memory_mb}MB available)")
        else:
            effective_workers = min(self.max_workers, len(task_ids))
        
        # Execute tasks in parallel with resource limits
        with ThreadPoolExecutor(max_workers=effective_workers) as executor:
            futures = {
                executor.submit(self._execute_single_task, tasks[task_id]): task_id 
                for task_id in task_ids
            }

            success = True
            for future in as_completed(futures):
                task_success = future.result()
                if not task_success:
                    success = False

            return success

    def _execute_single_task(self, task: BuildTask) -> bool:
        """Execute a single task with caching"""
        self.execution_stats['total_tasks'] += 1

        # Check cache first
        if self.cache.has_cached_result(task):
            if self.cache.restore_cached_result(task):
                self.execution_stats['cache_hits'] += 1
                return True

        # Execute task
        logger.info(f"→ {task.task_id} - executing")
        logger.debug(f"Command: {task.command}")
        start_time = time.time()

        try:
            # Ensure output directories exist
            for output in task.outputs:
                output_dir = os.path.dirname(output)
                if output_dir:
                    os.makedirs(output_dir, exist_ok=True)

            # Execute command using execution environment
            result = self.exec_env.execute(
                command=task.command,
                cwd=os.getcwd(),
                timeout=DEFAULT_TASK_TIMEOUT_SECONDS
            )

            execution_time = time.time() - start_time

            if result.returncode == 0:
                logger.info(f"✓ {task.task_id} ({execution_time:.1f}s) - completed")
                self.cache.cache_task_result(task, execution_time, True)
                self.execution_stats['cache_misses'] += 1
                return True
            else:
                logger.error(f"✗ {task.task_id} - failed: {result.stderr}")
                self.execution_stats['failed_tasks'] += 1
                return False

        except subprocess.TimeoutExpired:
            logger.error(f"✗ {task.task_id} - timeout after {DEFAULT_TASK_TIMEOUT_SECONDS}s")
            self.execution_stats['failed_tasks'] += 1
            return False
        except PermissionError as e:
            logger.error(f"✗ {task.task_id} - permission denied: {e}")
            self.execution_stats['failed_tasks'] += 1
            return False
        except OSError as e:
            logger.error(f"✗ {task.task_id} - OS error: {e}")
            self.execution_stats['failed_tasks'] += 1
            return False
        except Exception as e:
            logger.error(f"✗ {task.task_id} - unexpected error: {e}")
            self.execution_stats['failed_tasks'] += 1
            return False

    def _print_dry_run(self, graph: TaskGraph):
        """Print dry run information"""
        plan = graph.get_execution_plan()

        logger.info(f"Execution Plan:")
        logger.info(f"  Total tasks: {plan['total_tasks']}")
        logger.info(f"  Total stages: {plan['total_stages']}")
        logger.info(f"  Estimated time: {plan['estimated_time_seconds']:.1f}s")
        logger.info(f"  Max parallel tasks: {plan['max_parallel_tasks']}")

        for i, stage_info in enumerate(plan['stages']):
            stage_key = list(stage_info.keys())[0]
            tasks = stage_info[stage_key]
            logger.info(f"  {stage_key}: {tasks}")

    def _print_summary(self):
        """Print execution summary"""
        stats = self.execution_stats
        logger.info(f"\n✓ Build completed in {stats['total_time']:.1f}s")
        logger.info(f"✓ Tasks: {stats['total_tasks']} total, {stats['cache_hits']} cache hits, {stats['cache_misses']} executed")
        if stats['failed_tasks'] > 0:
            logger.error(f"✗ Failed tasks: {stats['failed_tasks']}")

File: src/executor.go
package main

import (
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// TaskExecutor executes tasks with caching, parallel execution, and resource-aware scheduling
type TaskExecutor struct {
	Cache               *BuildCache
	ChangeDetector      *ChangeDetector
	MaxWorkers          int
	MaxMemoryMB         int
	ExecEnv             *ExecutionEnvironment
	DisableCache        bool
	ExecutionStats      ExecutionStats
	BuildResult         *BuildResult            // Optional: for tracking execution in build report
	ShutdownManager     *ShutdownManager        // For graceful shutdown handling
	ResponseFileManager *ResponseFileManager    // For handling long command lines on Windows
	mu                  sync.Mutex
	pendingOutputs      map[string][]string     // Track outputs of in-progress tasks for cleanup
}

// ExecutionStats tracks execution statistics
type ExecutionStats struct {
	TotalTasks  int
	CacheHits   int
	CacheMisses int
	FailedTasks int
	TotalTime   float64
}

// NewTaskExecutor creates a new TaskExecutor
func NewTaskExecutor(cache *BuildCache, changeDetector *ChangeDetector, maxWorkers, maxMemoryMB int, execEnv *ExecutionEnvironment, disableCache bool) *TaskExecutor {
	if execEnv == nil {
		execEnv = NewNativeExecution()
	}

	// Initialize response file manager with cache directory
	cacheDir := ".buildy_cache"
	if cache != nil {
		cacheDir = cache.cacheDir
	}
	rfmConfig := DefaultResponseFileConfig(cacheDir)
	rfm := NewResponseFileManager(rfmConfig)

	te := &TaskExecutor{
		Cache:               cache,
		ChangeDetector:      changeDetector,
		MaxWorkers:          maxWorkers,
		MaxMemoryMB:         maxMemoryMB,
		ExecEnv:             execEnv,
		DisableCache:        disableCache,
		ShutdownManager:     GetShutdownManager(),
		ResponseFileManager: rfm,
		pendingOutputs:      make(map[string][]string),
		ExecutionStats: ExecutionStats{
			TotalTasks:  0,
			CacheHits:   0,
			CacheMisses: 0,
			FailedTasks: 0,
			TotalTime:   0,
		},
	}
	
	// Register cleanup for pending outputs and response files on shutdown
	if te.ShutdownManager != nil {
		te.ShutdownManager.RegisterCleanup(func() {
			te.cleanupPendingOutputs()
			if te.ResponseFileManager != nil {
				te.ResponseFileManager.Cleanup()
			}
		})
	}
	
	return te
}

// cleanupPendingOutputs removes partial outputs from interrupted tasks
func (te *TaskExecutor) cleanupPendingOutputs() {
	te.mu.Lock()
	defer te.mu.Unlock()
	
	for taskID, outputs := range te.pendingOutputs {
		log.Printf("Cleaning up partial outputs for interrupted task: %s", taskID)
		CleanupPartialOutputs(outputs)
	}
	te.pendingOutputs = make(map[string][]string)
}

// ExecuteTaskGraph executes all tasks in the graph
func (te *TaskExecutor) ExecuteTaskGraph(graph *TaskGraph, dryRun bool) bool {
	if len(graph.ExecutionStages) == 0 {
		if _, err := graph.BuildExecutionStages(); err != nil {
			log.Printf("ERROR: Failed to build execution stages: %v", err)
			return false
		}
	}

	startTime := time.Now()
	success := true
	completedTasks := 0
	totalTasks := len(graph.Tasks)

	log.Printf("Starting build with %d tasks in %d stages", totalTasks, len(graph.ExecutionStages))

	if dryRun {
		te.printDryRun(graph)
		return true
	}

	for stageNum, stageTasks := range graph.ExecutionStages {
		// Check for shutdown before starting each stage
		if te.ShutdownManager != nil && te.ShutdownManager.IsShuttingDown() {
			log.Printf("Build interrupted, stopping after stage %d", stageNum)
			success = false
			break
		}
		
		progress := 0.0
		if totalTasks > 0 {
			progress = (float64(completedTasks) / float64(totalTasks)) * 100
		}
		log.Printf("\nStage %d/%d: %d task(s) [%.1f%% complete]",
			stageNum+1, len(graph.ExecutionStages), len(stageTasks), progress)

		stageSuccess := te.executeStage(stageTasks, graph.Tasks)
		completedTasks += len(stageTasks)

		if !stageSuccess {
			success = false
			break
		}
	}

	te.ExecutionStats.TotalTime = time.Since(startTime).Seconds()
	te.printSummary()

	return success
}

// executeStage executes a single stage of tasks in parallel with resource-aware scheduling
func (te *TaskExecutor) executeStage(taskIDs []string, tasks map[string]*BuildTask) bool {
	if len(taskIDs) == 1 {
		// Single task - execute directly
		return te.executeSingleTask(tasks[taskIDs[0]])
	}

	// Resource-aware parallel execution
	// Calculate total memory requirements for the stage
	totalMemoryRequired := 0
	for _, taskID := range taskIDs {
		totalMemoryRequired += tasks[taskID].ResourceRequirements.MemoryMB
	}

	// Adjust max workers based on memory constraints
	effectiveWorkers := te.MaxWorkers
	if totalMemoryRequired > te.MaxMemoryMB {
		// Need to limit parallelism to avoid memory overload
		maxTaskMemory := 0
		for _, taskID := range taskIDs {
			if tasks[taskID].ResourceRequirements.MemoryMB > maxTaskMemory {
				maxTaskMemory = tasks[taskID].ResourceRequirements.MemoryMB
			}
		}
		effectiveWorkers = te.MaxMemoryMB / maxTaskMemory
		if effectiveWorkers < 1 {
			effectiveWorkers = 1
		}
		if effectiveWorkers > te.MaxWorkers {
			effectiveWorkers = te.MaxWorkers
		}
		if effectiveWorkers > len(taskIDs) {
			effectiveWorkers = len(taskIDs)
		}
		log.Printf("WARNING: Limiting parallelism to %d workers due to memory constraints "+
			"(%dMB required, %dMB available)", effectiveWorkers, totalMemoryRequired, te.MaxMemoryMB)
	} else {
		if effectiveWorkers > len(taskIDs) {
			effectiveWorkers = len(taskIDs)
		}
	}

	// Execute tasks in parallel with resource limits using goroutines
	var wg sync.WaitGroup
	taskChan := make(chan string, len(taskIDs))
	resultChan := make(chan bool, len(taskIDs))

	// Start worker goroutines
	for i := 0; i < effectiveWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for taskID := range taskChan {
				result := te.executeSingleTask(tasks[taskID])
				resultChan <- result
			}
		}()
	}

	// Send tasks to workers
	for _, taskID := range taskIDs {
		taskChan <- taskID
	}
	close(taskChan)

	// Wait for all workers to finish
	go func() {
		wg.Wait()
		close(resultChan)
	}()

	// Collect results
	success := true
	for result := range resultChan {
		if !result {
			success = false
		}
	}

	return success
}

// executeSingleTask executes a single task with caching
func (te *TaskExecutor) executeSingleTask(task *BuildTask) bool {
	// Check for shutdown before starting
	if te.ShutdownManager != nil && te.ShutdownManager.IsShuttingDown() {
		return false
	}
	
	te.mu.Lock()
	te.ExecutionStats.TotalTasks++
	te.mu.Unlock()

	// Update input hashes and cache key before cache lookup
	// This ensures cache keys are correct for tasks whose inputs were created by dependencies
	if err := task.UpdateInputHashesAndCacheKey(); err != nil {
		log.Printf("Warning: failed to update input hashes for %s: %v", task.TaskID, err)
	}

	// Check cache first (unless disabled by force flag)
	if !te.DisableCache && te.ChangeDetector != nil && te.ChangeDetector.IsCachedResultValid(task) {
		if err := te.Cache.RestoreCachedResult(task); err == nil {
			te.mu.Lock()
			te.ExecutionStats.CacheHits++
			if te.BuildResult != nil {
				te.BuildResult.RecordTaskExecution(task.TaskID, task.TaskType, 0, true, true)
			}
			te.mu.Unlock()
			return true
		}
		log.Printf("Cache restore failed for %s, falling back to execution", task.TaskID)
	}

	// Execute task
	log.Printf("→ %s - executing", task.TaskID)
	log.Printf("  Command: %s", task.Command)
	startTime := time.Now()

	// Ensure output directories exist
	for _, output := range task.Outputs {
		outputDir := filepath.Dir(output)
		if outputDir != "" {
			if err := os.MkdirAll(outputDir, 0755); err != nil {
				log.Printf("ERROR: Failed to create output directory %s: %v", outputDir, err)
				te.mu.Lock()
				te.ExecutionStats.FailedTasks++
				if te.BuildResult != nil {
					te.BuildResult.RecordTaskExecution(task.TaskID, task.TaskType, 0, false, false)
				}
				te.mu.Unlock()
				return false
			}
		}
	}

	// Track this task's outputs for cleanup on interruption
	te.mu.Lock()
	te.pendingOutputs[task.TaskID] = task.Outputs
	te.mu.Unlock()
	
	// Track task in shutdown manager
	if te.ShutdownManager != nil {
		te.ShutdownManager.TrackTask()
		defer te.ShutdownManager.TaskDone()
	}

	// Process command through response file manager for long commands (Windows)
	command := task.Command
	if te.ResponseFileManager != nil {
		processedCmd, err := te.ResponseFileManager.ProcessCommand(command, task.TaskID)
		if err != nil {
			log.Printf("WARNING: Failed to create response file for %s: %v", task.TaskID, err)
			// Continue with original command
		} else {
			command = processedCmd
		}
	}

	// Execute command using execution environment
	cwd, _ := os.Getwd()
	result, err := te.ExecEnv.Execute(command, cwd, DefaultTaskTimeoutSeconds)

	executionTime := time.Since(startTime).Seconds()

	// Remove from pending outputs tracking
	te.mu.Lock()
	delete(te.pendingOutputs, task.TaskID)
	te.mu.Unlock()

	if err != nil {
		log.Printf("✗ %s - failed: %v", task.TaskID, err)
		if result != nil && result.Stderr != "" {
			log.Printf("  stderr: %s", result.Stderr)
		}
		// Clean up partial outputs on failure
		CleanupPartialOutputs(task.Outputs)
		te.mu.Lock()
		te.ExecutionStats.FailedTasks++
		if te.BuildResult != nil {
			te.BuildResult.RecordTaskExecution(task.TaskID, task.TaskType, executionTime, false, false)
		}
		te.mu.Unlock()
		return false
	}

	if result.ReturnCode == 0 {
		log.Printf("✓ %s (%.1fs) - completed", task.TaskID, executionTime)
		te.Cache.CacheTaskResult(task, executionTime, true)
		te.mu.Lock()
		te.ExecutionStats.CacheMisses++
		if te.BuildResult != nil {
			te.BuildResult.RecordTaskExecution(task.TaskID, task.TaskType, executionTime, true, false)
		}
		te.mu.Unlock()
		return true
	}

	log.Printf("✗ %s - failed with exit code %d", task.TaskID, result.ReturnCode)
	if result.Stderr != "" {
		log.Printf("  stderr: %s", result.Stderr)
	}
	// Clean up partial outputs on failure
	CleanupPartialOutputs(task.Outputs)
	te.mu.Lock()
	te.ExecutionStats.FailedTasks++
	if te.BuildResult != nil {
		te.BuildResult.RecordTaskExecution(task.TaskID, task.TaskType, executionTime, false, false)
	}
	te.mu.Unlock()
	return false
}

// printDryRun prints dry run information
func (te *TaskExecutor) printDryRun(graph *TaskGraph) {
	plan, err := graph.GetExecutionPlan()
	if err != nil {
		log.Printf("ERROR: Failed to get execution plan: %v", err)
		return
	}

	log.Printf("Execution Plan:")
	log.Printf("  Total tasks: %v", plan["total_tasks"])
	log.Printf("  Total stages: %v", plan["total_stages"])
	log.Printf("  Estimated time: %.1fs", plan["estimated_time_seconds"])
	log.Printf("  Max parallel tasks: %v", plan["max_parallel_tasks"])

	if stages, ok := plan["stages"].([]map[string]any); ok {
		for _, stageInfo := range stages {
			for stageKey, tasksRaw := range stageInfo {
				if tasks, ok := tasksRaw.([]string); ok {
					log.Printf("  %s: %v", stageKey, tasks)
				}
			}
		}
	}
}

// printSummary prints execution summary
func (te *TaskExecutor) printSummary() {
	stats := te.ExecutionStats
	log.Printf("\n✓ Build completed in %.1fs", stats.TotalTime)
	log.Printf("✓ Tasks: %d total, %d cache hits, %d executed",
		stats.TotalTasks, stats.CacheHits, stats.CacheMisses)
	if stats.FailedTasks > 0 {
		log.Printf("✗ Failed tasks: %d", stats.FailedTasks)
	}
}

// GetExecutionStats returns the current execution statistics
func (te *TaskExecutor) GetExecutionStats() ExecutionStats {
	te.mu.Lock()
	defer te.mu.Unlock()
	return te.ExecutionStats
}
